### PR TITLE
enhancement(loki sink): Allow tenant_id to be templatable on loki sink

### LINKED ISSUE #5202: Allow `tenant_id` to be templatable on `loki` sink
Requested by user in discord: https://discord.com/channels/742820443487993987/746070604192415834/780815488774832158

They'd like to use the config:

```toml
[sinks.out]
  # General
  endpoint = "http://XXXX"
  inputs = ["my_transform_id"]
  type = "loki"
  labels.key = "value" # example
  labels.key = "{{ event_field }}" # example
  tenant_id = "{{ kubernetes.namespace }}"
```

It looks like this will require partitioned batching as tenant_id is sent per request.

https://github.com/timberio/vector/blob/18b9c0fade267e4833fa62067ba061ca84f35801/src/sinks/loki.rs#L187-L189

### PR DESCRIPTION
**Pull request by juchiast**
close #5202 

Signed-off-by: Duy Do <juchiast@gmail.com>

### CODE REVIEW

**Review by jszwedko — APPROVED** [review]
That was quick! Nice work @juchiast 

Do you think we could add a test for this behavior?

**Review by bruceg — APPROVED** [review]

File: src/sinks/loki.rs [hunk 1]
```diff
@@ -123,12 +124,30 @@ impl SinkConfig for LokiConfig {
...
-    type Input = LokiRecord;
-    type Output = serde_json::Value;
+    type Input = PartitionInnerBuffer<LokiRecord, PartitionKey>;
+    type Output = PartitionInnerBuffer<serde_json::Value, PartitionKey>;
 
     fn encode_event(&self, mut event: Event) -> Option<Self::Input> {
+        let tenant_id = self.tenant_id.as_ref().and_then(|t| {
+            t.render_string(&event)
+                .map_err(|missing| {
+                    error!(
```
[the remark this answers is not in the document]
Comment (juchiast) [reply src/sinks/loki.rs:141]: Let's add this in #4981 since there is no internal metrics for this source yet.

**Comment by juchiast** [comment]
@jszwedko  I added an integration test for this.

File: src/sinks/loki.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/sinks/loki.rs:141]: Sounds good :+1:

**Review by jszwedko — APPROVED** [review]
Looks good, thanks @juchiast

**Pull request merged**